app/src/components/jwtService.js

```diff
@@ -0,0 +1,105 @@
+const jose = require('jose');
+const config = require('config');
+const Problem = require('api-problem');
+
+const errorToProblem = require('./errorToProblem');
+
+const SERVICE = 'JwtService';
+
+const jwksUri = config.get('server.oidc.jwksUri');
+
+// Create a remote JWK set that fetches the JWK set from server with caching
+const JWKS = jose.createRemoteJWKSet(new URL(jwksUri));
+
+class JwtService {
+  constructor({ issuer, audience, maxTokenAge }) {
+    if (!issuer || !audience || !maxTokenAge) {
+      throw new Error('JwtService is not configured. Check configuration.');
+    }
+
+    this.audience = audience;
+    this.issuer = issuer;
+    this.maxTokenAge = maxTokenAge;
+  }
+
+  getBearerToken(req) {
+    if (req.headers && req.headers.authorization && req.headers.authorization.startsWith('Bearer ')) {
+      return req.headers.authorization.substring(7);
+    }
+    // do we want to throw errors?
+    return null;
+  }
+
+  async getTokenPayload(req) {
+    const bear = this.getBearerToken(req);
+    if (bear) {
+      return await this._verify(bear);
+    }
+    return null;
+  }
+
+  async _verify(token) {
+    // could throw JWTClaimValidationFailed
+    const { payload } = await jose.jwtVerify(token, JWKS, {
+      issuer: this.issuer,
+      audience: this.audience,
+      maxTokenAge: parseInt(this.maxTokenAge),
+    });
+    return payload;
+  }
+
+  async validateAccessToken(token) {
+    try {
+      await this._verify(token);
+      // these claims passed, just return true.
+      return true;
+    } catch (e) {
+      if (e instanceof jose.errors.JWTClaimValidationFailed) {
```

Comment:
The call to `errorToProblem` below is returning a `502` response when:
- The JWT header is invalid (`jose.errors.JWSInvalid`)
- The JWT signature is invalid (`jose.errors.JWSSignatureVerificationFailed`)
- The JWT is expired (`jose.errors.JWTExpired`)

There's a bunch of other errors that might happen. What's the the best approach here? Do we maybe want to just treat all JOSE errors as unauthenticated? Something like:

```suggestion
      if (e instanceof jose.errors.JOSEError) {
```
???

Or should we throw the `JOSEError`s and in the caller use the message in the response details? So many options!

Reply:
well, if JOSEError is a parent class, then perhaps we should catch that. Returning `false` is to simulate the behaviour of the `keycloak.grantManager` which just failed. No differentiation on why token was invalid or bad. So I am definitely good with catching `JOSEError` instead.

Reply:
I believe jwtVerify will only throw `JWTClaimValidationFailed` but we aren't really doing anything with claim field in the error anyway, so catching everything JOSE related should be fine.

Reply:
Looks great! Getting 401s for all of the above.